File: sdk/locator.go
package sdk

import (
	"context"
	"fmt"
	"strings"

	pb "github.com/ranjith035/morego/proto/v1"
)

type relativeConstraint struct {
	direction string
	target    *Locator
}

// Locator models visual elements in the device layout tree.
type Locator struct {
	session     *Session
	strategy    string
	selector    string
	index       int
	parent      *Locator
	constraints []relativeConstraint
}

func mapStrategy(s string) pb.LocatorStrategy {
	switch strings.ToUpper(s) {
	case "ACCESSIBILITY_ID":
		return pb.LocatorStrategy_LOCATOR_STRATEGY_ACCESSIBILITY_ID
	case "TEST_ID":
		return pb.LocatorStrategy_LOCATOR_STRATEGY_TEST_ID
	case "ROLE":
		return pb.LocatorStrategy_LOCATOR_STRATEGY_ROLE
	case "TEXT":
		return pb.LocatorStrategy_LOCATOR_STRATEGY_TEXT
	case "PLACEHOLDER":
		return pb.LocatorStrategy_LOCATOR_STRATEGY_PLACEHOLDER
	case "LABEL":
		return pb.LocatorStrategy_LOCATOR_STRATEGY_LABEL
	case "RESOURCE_ID":
		return pb.LocatorStrategy_LOCATOR_STRATEGY_RESOURCE_ID
	case "XPATH":
		return pb.LocatorStrategy_LOCATOR_STRATEGY_XPATH
	default:
		return pb.LocatorStrategy_LOCATOR_STRATEGY_UNSPECIFIED
	}
}

func mapDirection(d string) pb.RelativeDirection {
	switch strings.ToUpper(d) {
	case "ABOVE":
		return pb.RelativeDirection_RELATIVE_DIRECTION_ABOVE
	case "BELOW":
		return pb.RelativeDirection_RELATIVE_DIRECTION_BELOW
	case "LEFT_OF":
		return pb.RelativeDirection_RELATIVE_DIRECTION_LEFT_OF
	case "RIGHT_OF":
		return pb.RelativeDirection_RELATIVE_DIRECTION_RIGHT_OF
	default:
		return pb.RelativeDirection_RELATIVE_DIRECTION_UNSPECIFIED
	}
}

func (l *Locator) toProto() *pb.Locator {
	if l == nil {
		return nil
	}
	p := &pb.Locator{
		Strategy: mapStrategy(l.strategy),
		Selector: l.selector,
		Index:    int32(l.index),
	}
	if l.parent != nil {
		p.Parent = l.parent.toProto()
	}
	for _, c := range l.constraints {
		p.Constraints = append(p.Constraints, &pb.RelativeConstraint{
			Direction: mapDirection(c.direction),
			Target:    c.target.toProto(),
		})
	}
	return p
}

// Nth filters multiple selections by zero-indexed values.
func (l *Locator) Nth(index int) *Locator {
	return &Locator{
		session:     l.session,
		strategy:    l.strategy,
		selector:    l.selector,
		index:       index,
		parent:      l.parent,
		constraints: l.constraints,
	}
}

// Locator creates a chained sub-locator.
func (l *Locator) Locator(strategy string, selector string) *Locator {
	return &Locator{
		session:  l.session,
		strategy: strategy,
		selector: selector,
		parent:   l,
	}
}

// GetByText locates elements containing matching text inside this locator.
func (l *Locator) GetByText(text string) *Locator {
	return l.Locator("TEXT", text)
}

// GetByRole locates elements matching the class role.
func (l *Locator) GetByRole(role string) *Locator {
	return l.Locator("ROLE", role)
}

// GetByLabel locates elements matching the accessibility label.
func (l *Locator) GetByLabel(label string) *Locator {
	return l.Locator("LABEL", label)
}

// GetByPlaceholder locates elements matching the input placeholder attribute.
func (l *Locator) GetByPlaceholder(placeholder string) *Locator {
	return l.Locator("PLACEHOLDER", placeholder)
}

// GetByAccessibilityID locates elements by content description or accessibility ID.
func (l *Locator) GetByAccessibilityID(id string) *Locator {
	return l.Locator("ACCESSIBILITY_ID", id)
}

// GetByTestID locates elements by test identifier.
func (l *Locator) GetByTestID(id string) *Locator {
	return l.Locator("TEST_ID", id)
}

// Above adds a positional constraint that target must be above the reference locator.
func (l *Locator) Above(other *Locator) *Locator {
	l.constraints = append(l.constraints, relativeConstraint{
		direction: "ABOVE",
		target:    other,
	})
	return l
}

// Below adds a positional constraint that target must be below the reference locator.
func (l *Locator) Below(other *Locator) *Locator {
	l.constraints = append(l.constraints, relativeConstraint{
		direction: "BELOW",
		target:    other,
	})
	return l
}

// LeftOf adds a positional constraint that target must be left of the reference locator.
func (l *Locator) LeftOf(other *Locator) *Locator {
	l.constraints = append(l.constraints, relativeConstraint{
		direction: "LEFT_OF",
		target:    other,
	})
	return l
}

// RightOf adds a positional constraint that target must be right of the reference locator.
func (l *Locator) RightOf(other *Locator) *Locator {
	l.constraints = append(l.constraints, relativeConstraint{
		direction: "RIGHT_OF",
		target:    other,
	})
	return l
}

// Click locates and triggers touch clicks on the target locator.
func (l *Locator) Click(ctx context.Context) error {
	// 1. Resolve element via FindElement RPC
	findResp, err := l.session.device.driverClient.FindElement(ctx, &pb.FindElementRequest{
		SessionId: l.session.sessionID,
		Locator:   l.toProto(),
		TimeoutMs: 5000,
	})
	if err != nil {
		return fmt.Errorf("failed to locate element for Click: %w", err)
	}

	if findResp.Element == nil {
		return fmt.Errorf("element not found for locator: %+v", l)
	}

	// 2. Click the resolved element ID
	_, err = l.session.device.driverClient.Click(ctx, &pb.ClickRequest{
		DriverId: l.session.sessionID,
		Target: &pb.ClickRequest_ElementId{
			ElementId: findResp.Element.ElementId,
		},
	})
	if err != nil {
		return fmt.Errorf("click request failed: %w", err)
	}

	return nil
}

// Fill locates and enters text inside the target input element locator.
func (l *Locator) Fill(ctx context.Context, value string) error {
	// 1. Resolve element via FindElement RPC
	findResp, err := l.session.device.driverClient.FindElement(ctx, &pb.FindElementRequest{
		SessionId: l.session.sessionID,
		Locator:   l.toProto(),
		TimeoutMs: 5000,
	})
	if err != nil {
		return fmt.Errorf("failed to locate element for Fill: %w", err)
	}

	if findResp.Element == nil {
		return fmt.Errorf("element not found for locator: %+v", l)
	}

	// 2. Fill the resolved element ID
	_, err = l.session.device.driverClient.Fill(ctx, &pb.FillRequest{
		DriverId:  l.session.sessionID,
		ElementId: findResp.Element.ElementId,
		Value:     value,
	})
	if err != nil {
		return fmt.Errorf("fill request failed: %w", err)
	}

	return nil
}

File: sdk/typescript/src/locator.ts
export class Locator {
  private indexVal: number = 0;
  private parentVal?: Locator;
  private constraintsVal: Array<{ direction: string; target: Locator; distancePx?: number }> = [];

  constructor(
    private session: any,
    private strategy: string,
    private selector: string
  ) {}

  private toProto(): any {
    let strat = 'LOCATOR_STRATEGY_UNSPECIFIED';
    switch (this.strategy.toUpperCase()) {
      case 'ACCESSIBILITY_ID':
        strat = 'LOCATOR_STRATEGY_ACCESSIBILITY_ID';
        break;
      case 'TEST_ID':
        strat = 'LOCATOR_STRATEGY_TEST_ID';
        break;
      case 'ROLE':
        strat = 'LOCATOR_STRATEGY_ROLE';
        break;
      case 'TEXT':
        strat = 'LOCATOR_STRATEGY_TEXT';
        break;
      case 'PLACEHOLDER':
        strat = 'LOCATOR_STRATEGY_PLACEHOLDER';
        break;
      case 'LABEL':
        strat = 'LOCATOR_STRATEGY_LABEL';
        break;
      case 'RESOURCE_ID':
        strat = 'LOCATOR_STRATEGY_RESOURCE_ID';
        break;
      case 'XPATH':
        strat = 'LOCATOR_STRATEGY_XPATH';
        break;
    }

    const p: any = {
      strategy: strat,
      selector: this.selector,
      index: this.indexVal
    };

    if (this.parentVal) {
      p.parent = this.parentVal.toProto();
    }

    if (this.constraintsVal && this.constraintsVal.length > 0) {
      p.constraints = this.constraintsVal.map((c: any) => {
        let dir = 'RELATIVE_DIRECTION_UNSPECIFIED';
        switch (c.direction.toUpperCase()) {
          case 'ABOVE':
            dir = 'RELATIVE_DIRECTION_ABOVE';
            break;
          case 'BELOW':
            dir = 'RELATIVE_DIRECTION_BELOW';
            break;
          case 'LEFT_OF':
            dir = 'RELATIVE_DIRECTION_LEFT_OF';
            break;
          case 'RIGHT_OF':
            dir = 'RELATIVE_DIRECTION_RIGHT_OF';
            break;
        }
        return {
          direction: dir,
          target: c.target.toProto(),
          distance_px: c.distancePx || 0
        };
      });
    }

    return p;
  }

  private async resolve(): Promise<string> {
    return new Promise((resolve, reject) => {
      this.session.device.driverClient.FindElement(
        {
          session_id: this.session.sessionId,
          locator: this.toProto()
        },
        (err: any, response: any) => {
          if (err) return reject(err);
          if (!response.found) return reject(new Error(`Element not found: ${this.strategy}=${this.selector}`));
          resolve(response.element.element_id);
        }
      );
    });
  }

  nth(index: number): Locator {
    const loc = new Locator(this.session, this.strategy, this.selector);
    loc.indexVal = index;
    loc.parentVal = this.parentVal;
    loc.constraintsVal = [...this.constraintsVal];
    return loc;
  }

  locator(strategy: string, selector: string): Locator {
    const loc = new Locator(this.session, strategy, selector);
    loc.parentVal = this;
    return loc;
  }

  getByText(text: string): Locator {
    return this.locator('TEXT', text);
  }

  getByRole(role: string): Locator {
    return this.locator('ROLE', role);
  }

  getByLabel(label: string): Locator {
    return this.locator('LABEL', label);
  }

  getByPlaceholder(placeholder: string): Locator {
    return this.locator('PLACEHOLDER', placeholder);
  }

  getByAccessibilityID(id: string): Locator {
    return this.locator('ACCESSIBILITY_ID', id);
  }

  getByTestID(id: string): Locator {
    return this.locator('TEST_ID', id);
  }

  above(other: Locator, distancePx?: number): Locator {
    this.constraintsVal.push({ direction: 'ABOVE', target: other, distancePx });
    return this;
  }

  below(other: Locator, distancePx?: number): Locator {
    this.constraintsVal.push({ direction: 'BELOW', target: other, distancePx });
    return this;
  }

  leftOf(other: Locator, distancePx?: number): Locator {
    this.constraintsVal.push({ direction: 'LEFT_OF', target: other, distancePx });
    return this;
  }

  rightOf(other: Locator, distancePx?: number): Locator {
    this.constraintsVal.push({ direction: 'RIGHT_OF', target: other, distancePx });
    return this;
  }

  async click(): Promise<void> {
    const elId = await this.resolve();
    return new Promise((resolve, reject) => {
      this.session.device.driverClient.Click(
        { driver_id: this.session.sessionId, element_id: elId },
        (err: any) => {
          if (err) return reject(err);
          resolve();
        }
      );
    });
  }

  async fill(value: string): Promise<void> {
    const elId = await this.resolve();
    return new Promise((resolve, reject) => {
      this.session.device.driverClient.Fill(
        { driver_id: this.session.sessionId, element_id: elId, value },
        (err: any) => {
          if (err) return reject(err);
          resolve();
        }
      );
    });
  }
}
